modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -351,8 +369,7 @@ private CompletableFuture<Void> waitForActualState(long expirationTime) {
 
         return retryOperationUntilSuccess(raftClient::readIndex, e -> currentTimeMillis() > expirationTime, executor)
                 .orTimeout(timeout, TimeUnit.MILLISECONDS)
-                .thenCompose(storageIndexTracker::waitFor)
-                .thenRun(() -> waitForActualStateFuture.complete(null));
+                .thenCompose(idx -> storageIndexTracker.waitFor(idx));
```

Comment:
why have you moved completion of the future from here?

Reply:
Currently, the process of waiting for the actual state contains two parts:
1. waitForActualState
2. sendPrimaryReplicaChangeToReplicationGroup
The process is completed when the last stage is completed.
If we look at all places where the waitForActualState is used, you will see the sendPrimaryReplicaChangeToReplicationGroup is always used after.